### PR TITLE
added registration

### PR DESCRIPTION
**Pull request by TatianaPirch**
1.added registration;
2.added Role(entity, dao, service);
3.added UserDto;

### CODE REVIEW

File: src/main/java/mate/academy/spring/security/EmailValidator.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    private static final String EMAIL_PATTERN = "^[_A-Za-z0-9-+]+(.[_A-Za-z0-9-]+)*@"
+            + "[A-Za-z0-9-]+(.[A-Za-z0-9]+)*(.[A-Za-z]{2,})$";
+
+    @Override
+    public void initialize(ValidEmail constraintAnnotation) {
+    }
+
+    @Override
+    public boolean isValid(String email, ConstraintValidatorContext context) {
+        return (validateEmail(email));
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/security/EmailValidator.java]:
```suggestion
        return validateEmail(email);
```

File: src/main/java/mate/academy/spring/security/EmailValidator.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    private Matcher matcher;
+    private static final String EMAIL_PATTERN = "^[_A-Za-z0-9-+]+(.[_A-Za-z0-9-]+)*@"
+            + "[A-Za-z0-9-]+(.[A-Za-z0-9]+)*(.[A-Za-z]{2,})$";
+
+    @Override
+    public void initialize(ValidEmail constraintAnnotation) {
+    }
+
+    @Override
+    public boolean isValid(String email, ConstraintValidatorContext context) {
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/security/EmailValidator.java]: @TatianaPirch This is optional. Can you cover this with JUnit tests?

File: src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package mate.academy.spring.security;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import mate.academy.spring.dto.UserDto;
+import mate.academy.spring.security.annotations.PasswordMatches;
+
+public class PasswordMatchesValidator
+        implements ConstraintValidator<PasswordMatches, Object> {
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java]: @TatianaPirch Can we use like this?
```suggestion
        implements ConstraintValidator<PasswordMatches, UserDto> {
```

File: src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    @Override
+    public void initialize(PasswordMatches constraintAnnotation) {
+    }
+
+    @Override
+    public boolean isValid(Object obj, ConstraintValidatorContext context) {
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/security/PasswordMatchesValidator.java]: @TatianaPirch if we can parametrize class with UserDto, can we cover this method with JUnit? This is optional

File: src/main/java/mate/academy/spring/controller/RegistrationController.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    @GetMapping
+    public String registerForm(Model model) {
+        model.addAttribute("userDto", new UserDto());
+        return "registration";
+    }
+
+    @PostMapping
+    public String registerUser(@ModelAttribute @Valid UserDto userDto,
+            BindingResult result, Model model) {
+        User newUser = new User(userDto.getUsername(), userDto.getPassword(),
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/controller/RegistrationController.java]: it's better to use DtoUtil.convertDto. In this case, you will have the converting login in one place (DtoUtil class).

File: src/main/java/mate/academy/spring/security/EmailValidator.java (hunk 2 above)
Comment (TatianaPirch) [reply src/main/java/mate/academy/spring/security/EmailValidator.java]: Today, my favorite topic is multithreading :)
I will try to do it during the weekend. Ok?

Comment (boroda4436) [reply src/main/java/mate/academy/spring/security/EmailValidator.java]: np)

File: src/main/java/mate/academy/spring/dto/DtoUtil.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package mate.academy.spring.dto;
+
+import mate.academy.spring.entity.User;
+import org.springframework.stereotype.Component;
+
+@Component
```
Comment (boroda4436) [inline src/main/java/mate/academy/spring/dto/DtoUtil.java]: @TatianaPirch Usually util classes are not components) but let we leave it as is

**Review by boroda4436 — APPROVED** [review]

**Pull request merged**